core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1585,8 +1597,17 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
 
   private def acls(path: String): Seq[ACL] = ZkData.defaultAcls(isSecure, path)
 
-  private def retryRequestUntilConnected[Req <: AsyncRequest](request: Req): Req#Response = {
-    retryRequestsUntilConnected(Seq(request)).head
+  private def retryRequestUntilConnected[Req <: AsyncRequest](request: Req, expectedControllerZkVersion: Int = ZkVersion.MatchAnyVersion): Req#Response = {
+    retryRequestsUntilConnected(Seq(request), expectedControllerZkVersion).head
+  }
+
+  private def retryRequestsUntilConnected[Req <: AsyncRequest](requests: Seq[Req], expectedControllerZkVersion: Int): Seq[Req#Response] = {
+    if (expectedControllerZkVersion < KafkaController.InitialControllerEpoch) {
```

Comment:
The code here is comparing the zkversion with epoch, which seems misleading.

Also, `expectedControllerZkVersion < 0`, will `expectedControllerZkVersion` be anything other than `MatchAnyVersion`? If not, it seems better to explicitly check `expectedControllerZkVersion == ZkVersion.MatchAnyVersion`. And if they are not equal, we can throw IllegalStateException if `expectedControllerZkVersion` is negative.

Reply:
Fixed.